File: frontend/api/medium.ts
import type { AxiosInstance } from "axios";
import { objectToFormData } from "~/lib/helpers/formdata";

export interface CreateMediumComponentRelationInput {
  component: {
    id: number | null;
    component_formula: string;
    component_molar_mass: number;
    type_id: number;
  };
  mg_per_liter: string;
}

export interface CreateMediumComponentRelationInputFormatted {
  component_id: number | null;
  mg_per_liter: string;
}
export interface CreateMediumInput {
  name: string;
  description: string;
  thumbnail: File | null;
  components: CreateMediumComponentRelationInput[];
}
export interface CreateMediumInputFormatted {
  name: string;
  description: string;
  thumbnail: File | null;
  components: CreateMediumComponentRelationInputFormatted[];
}

export interface MediumFull {
  id: number;
  name: string;
  description: string;
  thumbnail: string;
  components: MediumComponent[];
}

export interface MediumComponent {
  component: {
    id: number|null;
    type_id: number;
    component_formula: string;
    component_molar_mass: number;
  };
  mg_per_liter: string;
}

export class Medium {
  instance: AxiosInstance;
  constructor(instance: AxiosInstance) {
    this.instance = instance;
  }
  async createMedium(payload: CreateMediumInputFormatted) {
    const data = objectToFormData(payload);
    return this.instance.post("/mediums", data);
  }
  async getMediums() {
    return this.instance.get<MediumFull[]>("/mediums");
  }
}

File: golang-backend/models/medium.go
package models

type Medium struct {
  Id          int                `json:"id" form:"id"`
  Name        string             `json:"name" form:"name"`
  Description string             `json:"description" form:"description"`
  Thumbnail   string             `json:"thumbnail" form:"thumbnail"`
  Components  []Medium_Component `json:"components" form:"components"`
}

type MediumWithComponents struct {
	Id          int                `json:"id"`
	Name        string             `json:"name"`
	Description string             `json:"description"`
	Thumbnail   string             `json:"thumbnail"`
	Components  []Medium_ComponentFull `json:"components"`
}
